internal/command/command.go

```diff
@@ -355,10 +356,246 @@ var CmdUpdateApis = &Command{
 			return nil
 		}
 
-		return push(ctx, languageRepo, startOfRun)
+		return push(ctx, languageRepo, startOfRun, "", "")
 	},
 }
 
+var CmdCreateReleasePR = &Command{
+	Name:  "create-release-pr",
+	Short: "Generate a PR for release",
+	Run: func(ctx context.Context) error {
+		err := checkFlags()
+		if err != nil {
+			return err
+		}
+		languageRepo, repoPath, inputDirectory, err := setupLanguageAndInputFolders(ctx)
+		if err != nil {
+			return err
+		}
+
+		pipelineState, err := loadState(languageRepo)
+		if err != nil {
+			fmt.Println("Error loading pipeline state:", err)
+			return err
+		}
+
+		librariesToRelease, prDescription, err := generateReleaseCommitForEachLibrary(ctx, repoPath, languageRepo, inputDirectory, pipelineState)
+		if err != nil {
+			return err
+		}
+
+		return generateReleasePr(ctx, librariesToRelease, pipelineState, repoPath, languageRepo, prDescription)
+	},
+}
+
+func setupLanguageAndInputFolders(ctx context.Context) (*gitrepo.Repo, string, string, error) {
+	startOfRun := time.Now()
+	tmpRoot, err := createTmpWorkingRoot(startOfRun)
+	var languageRepo *gitrepo.Repo
+	if flagRepoRoot == "" {
+		languageRepo, err = cloneLanguageRepo(ctx, flagLanguage, tmpRoot)
+		if err != nil {
+			return nil, "", "", err
+		}
+	} else {
+		languageRepo, err = gitrepo.Open(ctx, flagRepoRoot)
+		if err != nil {
+			return nil, "", "", err
+		}
+	}
+
+	inputDir := filepath.Join(tmpRoot, "inputs")
+	if err := os.Mkdir(inputDir, 0755); err != nil {
+		fmt.Println("Failed to create input directory:", err)
+		return nil, "", "", err
+	}
+
+	repoPath := flagRepoRoot
+	if repoPath == "" {
+		repoPath = filepath.Join(tmpRoot, fmt.Sprintf("google-cloud-%s", flagLanguage))
+	}
+
+	return languageRepo, repoPath, inputDir, nil
+}
+
+func checkFlags() error {
+	if !supportedLanguages[flagLanguage] {
+		return fmt.Errorf("invalid -language flag specified: %q", flagLanguage)
+	}
+	if flagImage == "" {
+		return fmt.Errorf("Create release pr requires image")
+	}
+	return nil
+}
+
+func generateReleasePr(ctx context.Context, librariesToRelease map[string]string, pipelineState *statepb.PipelineState, repoPath string, repo *gitrepo.Repo, prDescription string) error {
+	if len(librariesToRelease) > 0 {
+		updateLibraryMetadata(librariesToRelease, pipelineState, repoPath)
```

Comment:
I think we should update the pipeline state on disk as we create the other commits - so that each commit is self-contained. That way:

- If we revert a commit due to testing failing (in a later change) we stay consistent
- Each library release commit is consistent in its versioning, so it doesn't matter what the release build uses to read the version (it could use the pipeline state or the in-language version)
- If the PR is reviewed manually, a maintainer can drop any commit for a library they don't want to release, and the PR itself still remains valid.